lib/spack/spack/database.py

```diff
@@ -339,7 +365,141 @@ def read_transaction(self):
         """Get a read lock context manager for use in a `with` block."""
         return ReadTransaction(self.lock, acquire=self._read)
 
-    def prefix_lock(self, spec):
+    def _failed_spec_path(self, spec):
+        """Return the path to the spec's failure file, which may not exist."""
+        if not spec.concrete:
+            raise ValueError('Concrete spec required for failure path for {0}'
+                             .format(spec.name))
+
+        return os.path.join(self._failure_dir,
+                            '{0}-{1}'.format(spec.name, spec.full_hash()))
+
+    def clear_failure(self, spec, force=False):
+        """Remove any persistent and cached failure tracking for the spec."""
+        if self.prefix_failure_locked(spec):
+            if not force:
+                tty.log('Retaining failure marking for {0} due to lock'
+                        .format(spec.name))
+                return
+            else:
+                tty.warn('Removing failure marking despite lock for {0}'
+                         .format(spec.name))
+
+        if spec.prefix in self._prefix_failures:
+            del self._prefix_failures[spec.prefix]
```

Comment:
If I understand the code correctly, `self._prefix_failures` contains a mapping from the prefix of the failed spec to the lock acquired to mark the failure. Wouldn't it be safer then to release the lock before deleting it? Am I missing some context that will trigger equivalent calls?

Reply:
Excellent point!